### ISSUE TITLE
<pad_token> doesn't seem to be set correctly!

### ISSUE THREAD

**Issue by piaolaidelangman** [QwenLM/Qwen2#814]
# The attention mask which is auto generated by transformers generate function is not the same as what the tokenizer generated! 

# This will lead to unexpected behavior!

# Reproduce:
``` python
from transformers import AutoModelForCausalLM, AutoTokenizer
qwen_model_name = "Qwen/Qwen2-0.5B-Instruct"
device = "cuda"

model = AutoModelForCausalLM.from_pretrained(
    qwen_model_name,
    device_map=device,
    trust_remote_code=True
)
tokenizer = AutoTokenizer.from_pretrained(qwen_model_name)

prompts = ['Question 1, what can you do?', "Question 2, Who is the most powerful man in the world?"]
inputs = tokenizer(prompts, return_tensors="pt", padding=True)
print(inputs['attention_mask'])
print(inputs['input_ids'])

generate_ids = model.generate(inputs.input_ids.to(device), max_length=30, tokenizer=tokenizer)
res = tokenizer.batch_decode(generate_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False)
print(res)

```

The tokenizer generated attention_mask is :
```
tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])
```
and the input_ids is:
```
tensor([[ 14582,    220,     16,     11,   1128,    646,    498,    653,     30,
         151643, 151643, 151643, 151643, 151643],
        [ 14582,    220,     17,     11,  10479,    374,    279,   1429,   7988,
            883,    304,    279,   1879,     30]])
```
The tokenizer use <|endoftext|> which ids is 151643 as the pad_token.

However, if we don't set the attention_mask in the generate function, [the generate function will prepare the attention mask itself](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L1728).

After executing [this function](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L475), the attention mask is:
```
{'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]), 'use_cache': True}
```

Undoubtedly, this process will use <pad_token>, however, the [model config file](https://huggingface.co/Qwen/Qwen2-0.5B-Instruct/blob/main/config.json) does not contain the pad_token_id, then transformers use [<eos_token> as <pad_token>](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L1542), which idx is 151645.

**Comment by jklj077** [comment]
Hi, I think that it is related to batch generation.
1. for batch generation, please enable left padding
2. the example code in the README does pass attention_mask to model.generate
3. the `_prepare_attention_mask_for_generation` uses the generation config, where the pad_token_id is set.
